<a href="https://colab.research.google.com/github/Jose-Gabriel-Rodriguez/MachineLearning/blob/main/Unidad3/Practica_1_KNN_Frutas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![image](https://github.com/JoseGabriel-ITD/Probabilidad-y-Estadistica/blob/main/Cintilla2004.png?raw=true)

# **Machine Learning y Deep Learning**

## Unidad 3

### Practica 1: **Algortimo K-NN, los K-Vecinos mas Cercanos**

### Facilitador: *Dr. José Gabriel Rodríguez Rivas*

### Alumno:  _____________


## **Practica 1: Algortimo K-NN, los K-Vecinos mas Cercanos**


### **Funcionamiento**
Dado un conjunto de entrenamiento **X_train** con etiquetas **y_train** y dada una nueva instancia x_test se va a clasificar:

1. Encontar las instancias más similares (llamémoslas X_NN) a x_test que están en X_train.
2. Obtenga las etiquetas y_NN para las instancias de X_NN
3. Prediga la etiqueta para x_test combinando las etiquetas y_NN por ejemplo, mayoría simple de votos

**Un algoritmo de K-vecinos más cercanos necesita cuatro cosas especificas**
1. Una métrica de distancia
2. ¿Cuántos vecinos "más cercanos" hay que mirar?
3. Función de ponderación opcional en los puntos vecinos
4. Método para agregar las clases de puntos vecinos



## **Fruit dataset**


- Es uno de los conjuntos de datos más conocidos en el mundo del aprendizaje automático y la ciencia de datos.

- Consta de 59 registros con 4 tipos de frutas



In [2]:
# Importar librerias

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, confusion_matrix

In [3]:
frutas = pd.read_csv("frutas.csv")
frutas.head()

fruit_label fruit_name fruit_subtype  mass  width  height  color_score
0            1      apple  granny_smith   192    8.4     7.3         0.55
1            1      apple  granny_smith   180    8.0     6.8         0.59
2            1      apple  granny_smith   176    7.4     7.2         0.60
3            2   mandarin      mandarin    86    6.2     4.7         0.80
4            2   mandarin      mandarin    84    6.0     4.6         0.79

## **Modelo de clasificación usando parametro k = 1, es decir, 1 vecino**


### **Selección de variables predictoras (x) y variable a predecir (y)**

In [4]:
# Seleccion de caracteristicas o variables predictoras o variables independientes
X = frutas[["height", "width", "mass", "color_score"]]

# Variable objetivo o variable a predecir
y = frutas["fruit_name"]

### **Dividir dataset en conjunto de entrenamiento y prueba**

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

### **Entrenamiento del modelo**

- Entrenar un modelo se refiere al proceso de ajustar un modelo a los datos, de modo que el modelo pueda aprender a hacer predicciones basadas en los datos.
- En el aprendizaje automático, esto normalmente se hace optimizando los parámetros del modelo para minimizar el error entre las predicciones del modelo y los valores reales.

In [6]:
#Definimos el clasificador KNN , definimos 1 vecinos
clasificador_KNN = KNeighborsClassifier(n_neighbors= 1)

# realizar el entrenamiento con las variables definidas
clasificador_KNN.fit(X_train, y_train)
print("Precision del clasificador en fase de entrenamiento",
      clasificador_KNN.score(X_train, y_train) )

Precision del clasificador en fase de entrenamiento 1.0


### **Prueba del Modelo**
- La prueba del modelo implica **evaluar su desempeño en un conjunto de datos no visto previamente (el conjunto de prueba)** para medir qué tan bien generaliza al trabajar con datos nuevos.
- Este proceso es fundamental para verificar la utilidad práctica del modelo y prevenir problemas como el sobreajuste o subajuste.  

### **Generalización**
- La capacidad del modelo para generalizar es su habilidad para hacer predicciones precisas en datos nuevos. Un modelo bien probado debe demostrar que no solo memoriza los datos de entrenamiento, sino que entiende los patrones subyacentes.
- **Sobreajuste (Overfitting)**: El modelo aprende demasiado bien los detalles y ruido de los datos de entrenamiento, lo que lleva a un mal desempeño en el conjunto de prueba.
- **Subajuste (Underfitting)**: El modelo no captura suficientemente la relación entre los datos, llevando a un bajo desempeño en ambos conjuntos.

In [7]:
# Realizar una prediccion con los datos de prueba
y_pred = clasificador_KNN.predict(X_test)

# Crear un informe de texto que muestre las principales métricas de clasificación.
print("Reporte de métricas del clasificador: \n", classification_report(y_test, y_pred))

# Coeficiente de determinación
print( "Precisión:", clasificador_KNN.score(X_test, y_test) )

Reporte de métricas del clasificador: 
               precision    recall  f1-score   support

       apple       0.00      0.00      0.00         4
       lemon       1.00      0.60      0.75         5
    mandarin       1.00      1.00      1.00         2
      orange       0.38      0.75      0.50         4

    accuracy                           0.53        15
   macro avg       0.59      0.59      0.56        15
weighted avg       0.57      0.53      0.52        15

Precisión: 0.5333333333333333


### **Reporte de los resultados**

Qué significa cada métrica:
* **Precision (precisión)**: De todas las veces que el modelo predijo una fruta, ¿cuántas veces acertó?  
* **Recall (sensibilidad)**: De todas las frutas reales de una clase, ¿cuántas fueron correctamente detectadas?  
* **F1-score**: Media armónica entre precisión y recall → equilibrio entre ambos.  
* **Support**: Número de ejemplos de cada clase en el conjunto de prueba.

**Accuracy = 0.5333** → de 15 ejemplos de prueba, el modelo acertó 8 (0.5333 × 15 ≈ 8).

#### **apple**
- **Precision = 0.0:** El 0% de las predicciones como "apple" fueron correctas. Es decir, el modelo confundió otras clases como "apple" en el 100% de los casos.
- **Recall = 0.0**: El modelo identificó correctamente el 0% de las "apple" reales en los datos.
- **F1-score = 0.0**: Equilibrio entre precisión y recall. Esto indica que el modelo tiene un desempeño nulo en esta clase.
- **Support** = 4: Había 4 ejemplos de "apple" reales en los dato

#### **lemon**
- **Precision = 1.00**: Todas las frutas predichas como "lemon" eran realmente "lemon". El modelo no generó falsos positivos para esta clase.
- **Recall = 0.60**: El modelo identificó solo el 60% de los "lemon" reales. (falsos negativos).
- **F1-score = 0.75**: Un equilibrio aceptable entre precisión (muy alta) y recall (moderado), pero limitado por el bajo recall.
- **Support = 5**: Había 5 ejemplos de "lemon" reales.

#### **mandarin**
- **Precision = 1.00**: Todas las frutas predichas como "mandarina" eran efectivamente "mandarina".
- **Recall = 1.00**: El modelo identificó todas las "mandarinas" reales.
- **F1-score = 1.00**: Desempeño perfecto en esta clase.
- **Support = 2**: Había solo 2 muestras de "mandarinas", lo cual facilita obtener métricas perfectas (aunque no garantiza que el modelo sea robusto para esta clase).

#### **orange**
- **Precision = 0.38**: El 38% de las predicciones como "orange" eran correctas. Hubo un 60% de falsos positivos.
- **Recall = 0.75**: El modelo identificó el 75% de las "orange" reales, pero omitió el 25% (falsos negativos).
- **F1-score = 0.50**: Desempeño bajo
- **Support = 4**: Había 4 ejemplos de "orange" reales.

### **Desempeño General**
- **Accuracy = 0.53 (53%)**: El modelo predijo correctamente el 53% de todas las frutas en el conjunto de prueba.
- **Macro Average (Promedio Macro)**: Calcula el promedio de precisión, recall y F1-score dando el mismo peso a cada clase, sin importar su tamaño.
- **Weighted Average (Promedio Ponderado)**: Calcuula el Promedio ponderado por el tamaño de cada clase

### **Concluciones**
* El modelo no acierta ninguna manzana (apple TP = 0).
* Algunas clases (lemon y mandarin) están muy bien identificadas;
* apple y orange se confunden entre sí (o con otras clases).

### **Consideraciones**
- Un buen modelo busca alta precisión y recall en ambas clases, especialmente si las clases son igualmente importantes.
- Si hay un desbalance entre las clases (por ejemplo, muchas más naranjas que manzanas), las métricas como el F1-score son más representativas que solo la precisión.

### **Matriz de confusión**

La matriz de confusión muestra cómo el modelo clasificó las instancias reales en diferentes clases. Las filas representan las clases reales y las columnas las predicciones:

- Diagonal principal: Son las predicciones correctas (True Positives, TP).
- Fuera de la diagonal: Son las predicciones incorrectas:
     - Falsos positivos (FP): Instancias incorrectamente asignadas a una clase.
     - Falsos negativos (FN): Instancias de una clase que no fueron identificadas correctamente.

In [8]:
# Resumen de las predicciones realizadas por el clasificador
print ("Matriz de confusión:\n", confusion_matrix(y_test, y_pred), "\n")

Matriz de confusión:
 [[0 0 0 4]
 [1 3 0 1]
 [0 0 2 0]
 [1 0 0 3]] 



### **Interpretación**

| Fruta real               | Predicciones correctas | Predicciones incorrectas                      | Observaciones                                                                                                                                        |
| ------------------------ | ---------------------- | --------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------- |
| **apple** (manzana)      | 0/4 correctas        | Todas fueron clasificadas como **orange (4)** | El modelo no logra distinguir manzanas, probablemente se parecen mucho en características físicas (masa, altura, ancho, color_score) a las naranjas. |
| **lemon** (limón)        | 3/5 correctas        | 1 como **apple**, 1 como **orange**           | Lo hace bien, pero hay algo de confusión con frutas de color similar.                                                                                |
| **mandarin** (mandarina) | 2/2 correctas         | Ningún error                                  | Muy buen desempeño, los datos de esta clase están bien separados.                                                                                    |
| **orange** (naranja)     | 3/4 correctas        | 1 mal clasificada como **apple**              | El modelo confunde ocasionalmente naranjas con manzanas.                                                                                             |


### **Calcular la importancia de características (Permutation Importance)**

* Este método evalúa cuánto empeora el rendimiento del modelo cuando se altera aleatoriamente una característica.
* Mide cuánto baja la precisión del modelo si se altera una variable.
Cuanto mayor sea esa caída, más importante es la variable.

In [9]:
from sklearn.inspection import permutation_importance

result = permutation_importance(clasificador_KNN, X_test, y_test, n_repeats=30, random_state=42)

importancia = pd.DataFrame({
    "Variable": X.columns,
    "Importancia Media": result.importances_mean,
    "Desviación": result.importances_std
}).sort_values("Importancia Media", ascending=False)

print(importancia)

      Variable  Importancia Media  Desviación
2         mass           0.266667    0.095839
3  color_score          -0.004444    0.016630
0       height          -0.028889    0.063674
1        width          -0.062222    0.048483


In [10]:
# Visualizar importancias con Plotly
import plotly.express as px
fig4 = px.bar(importancia, x="Variable", y="Importancia Media",
              error_y="Desviación", title="Importancia de características (KNN - Permutation Importance)",
              text_auto=".3f", color="Variable")
fig4.update_layout(width=600, height=400)
fig4.show()

## **Modelo de clasificación usando parametro k = 3, es decir, 3 vecinos**

In [11]:
#Definimos el clasificador KNN , definimos 3 vecinos
clasificador_KNN3 = KNeighborsClassifier(n_neighbors= 3 )

# realizar el entrenamiento con las variables definidas
clasificador_KNN3.fit(X_train, y_train)
print("Precision del clasificador en fase de entrenamiento",
      clasificador_KNN3.score(X_train, y_train) )

Precision del clasificador en fase de entrenamiento 0.8409090909090909


In [12]:
# Realizar una prediccion con los datos de prueba
y_pred = clasificador_KNN3.predict(X_test)

# Crear un informe de texto que muestre las principales métricas de clasificación.
print("Reporte de métricas del clasificador: \n", classification_report(y_test, y_pred))

# Coeficiente de determinación
print( "Precisión:", clasificador_KNN3.score(X_test, y_test) )

Reporte de métricas del clasificador: 
               precision    recall  f1-score   support

       apple       0.67      0.50      0.57         4
       lemon       0.50      0.40      0.44         5
    mandarin       1.00      1.00      1.00         2
      orange       0.50      0.75      0.60         4

    accuracy                           0.60        15
   macro avg       0.67      0.66      0.65        15
weighted avg       0.61      0.60      0.59        15

Precisión: 0.6


In [13]:
# Resumen de las predicciones realizadas por el clasificador
print ("Matriz de confusión:\n", confusion_matrix(y_test, y_pred), "\n")

Matriz de confusión:
 [[2 1 0 1]
 [1 2 0 2]
 [0 0 2 0]
 [0 1 0 3]] 



### **Interpretación de resultados**  

**Matriz de confusión**  
| Fruta real                  | Predicciones correctas | Errores                         | Observación                                                                         |
| --------------------------- | ---------------------- | ------------------------------- | ----------------------------------------------------------------------------------- |
| **Apple (manzana)**      | 2 de 4 correctas       | 1 como *lemon*, 1 como *orange* | Mejoró respecto a k=1 (antes 0 aciertos). Ahora logra identificar algunas manzanas. |
| **Lemon (limón)**        | 2 de 5 correctas       | 1 como *apple*, 2 como *orange* | Sigue habiendo confusión, especialmente con *orange*.                               |
| **Mandarin (mandarina)** | 2 de 2 correctas       | 0 errores                       | Perfecta, igual que antes.                                                          |
| **Orange (naranja)**     | 3 de 4 correctas       | 1 como *lemon*                  | Buen desempeño, pero con cierta confusión con limones.                              |  

**Reporte de clasificación:**  


| Clase        | Precision | Recall | F1-score | Support | Interpretación                                                                                                          |
| ------------ | --------- | ------ | -------- | ------- | ----------------------------------------------------------------------------------------------------------------------- |
| **apple**    | 0.67      | 0.50   | 0.57     | 4       | De las frutas predichas como *apple*, el 67% eran realmente manzanas, y logró detectar la mitad de las manzanas reales. |
| **lemon**    | 0.50      | 0.40   | 0.44     | 5       | Metade de los limones reales fueron correctamente identificados, pero el modelo a veces los clasifica como *orange*.    |
| **mandarin** | 1.00      | 1.00   | 1.00     | 2       | Separación clara y bien definida.                                                                                       |
| **orange**   | 0.50      | 0.75   | 0.60     | 4       | Predice bien la mayoría de las naranjas, aunque confunde algunos limones como naranjas.                                 |  

**Metricas Globales**


| Métrica global      | Valor | Interpretación                           |
| ------------------- | ----- | ---------------------------------------- |
| **Accuracy**        | 0.60  | Mejoró de 0.53 → 0.60                   |
| **Macro avg F1**    | 0.65  | Promedio balanceado, más estable         |
| **Weighted avg F1** | 0.59  | Pondera por tamaño de clase, mejora leve |


## **Modelo de clasificación usando parametro k = 5, es decir, 5 vecinos**

In [14]:
#Definimos el clasificador KNN , definimos 5 vecinos
clasificador_KNN5 = KNeighborsClassifier(n_neighbors= 5 )

# realizar el entrenamiento con las variables definidas
clasificador_KNN5.fit(X_train, y_train)
print("Precision del clasificador en fase de entrenamiento",
      clasificador_KNN5.score(X_train, y_train) )

Precision del clasificador en fase de entrenamiento 0.8181818181818182


In [15]:
# Realizar una prediccion con los datos de prueba
y_pred = clasificador_KNN5.predict(X_test)

# Crear un informe de texto que muestre las principales métricas de clasificación.
print("Reporte de métricas del clasificador: \n", classification_report(y_test, y_pred))

# Coeficiente de determinación
print( "Precisión:", clasificador_KNN5.score(X_test, y_test) )

Reporte de métricas del clasificador: 
               precision    recall  f1-score   support

       apple       0.67      0.50      0.57         4
       lemon       0.50      0.40      0.44         5
    mandarin       1.00      1.00      1.00         2
      orange       0.50      0.75      0.60         4

    accuracy                           0.60        15
   macro avg       0.67      0.66      0.65        15
weighted avg       0.61      0.60      0.59        15

Precisión: 0.6


In [16]:
# Resumen de las predicciones realizadas por el clasificador
print ("Matriz de confusión:\n", confusion_matrix(y_test, y_pred), "\n")

Matriz de confusión:
 [[2 1 0 1]
 [1 2 0 2]
 [0 0 2 0]
 [0 1 0 3]] 



### **Interpretación**

* **Los resultados son similares a los obtenidos con k=3.**

## **Probar distintos valores de k**

In [17]:
for k in range(1, 20):
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, y_train)
    print(f"k={k} → Precisión: {knn.score(X_test, y_test):.2f}")


k=1 → Precisión: 0.53
k=2 → Precisión: 0.53
k=3 → Precisión: 0.60
k=4 → Precisión: 0.60
k=5 → Precisión: 0.60
k=6 → Precisión: 0.40
k=7 → Precisión: 0.60
k=8 → Precisión: 0.53
k=9 → Precisión: 0.53
k=10 → Precisión: 0.53
k=11 → Precisión: 0.47
k=12 → Precisión: 0.47
k=13 → Precisión: 0.47
k=14 → Precisión: 0.47
k=15 → Precisión: 0.40
k=16 → Precisión: 0.40
k=17 → Precisión: 0.40
k=18 → Precisión: 0.47
k=19 → Precisión: 0.40


### **Interpretacion de estos resultados**

**k muy pequeño (1–2)**
* El modelo se sobreajusta (overfitting):
* Clasifica casi idénticamente a los ejemplos de entrenamiento.  

**k intermedio (3–5)**
* El modelo alcanza su máxima estabilidad y generalización.
* Precisión más alta (0.60).

**k grande (>6)**
* El modelo empieza a subajustar (underfitting):
* Promedia demasiados vecinos.
* Pierde la capacidad de distinguir clases con pocos datos (como mandarin o apple).
* Precisión baja (0.40–0.47).

### **Otras alternativas para Intentar mejorar el rendimiento**

* Probar distintas **métricas de distancia**.
* Por defecto KNN usa distancia Euclidiana, pero se puede cambiar, por ejemplo: **metric='manhattan'**.
* Ajustar la normalización de datos


In [31]:
from sklearn.preprocessing import StandardScaler

X = frutas[["height", "width", "mass", "color_score"]]
y = frutas["fruit_name"]

# Dividir primero, stratify es para que mantenga la misma proporción de clases en ambos conjuntos
X_train, X_test, y_train, y_test = train_test_split( X, y,  random_state=42, stratify=y)

# Normalizar variables (KNN es sensible a la escala) y Escalar correctamente
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Entrenar KNN
knn_scaled = KNeighborsClassifier(n_neighbors=3)
knn_scaled.fit(X_train_scaled, y_train)

y_pred_scaled = knn_scaled.predict(X_test_scaled)
print("\nReporte de clasificación:")
print(classification_report(y_test, y_pred_scaled))


Reporte de clasificación:
              precision    recall  f1-score   support

       apple       1.00      0.80      0.89         5
       lemon       1.00      1.00      1.00         4
    mandarin       1.00      1.00      1.00         1
      orange       0.83      1.00      0.91         5

    accuracy                           0.93        15
   macro avg       0.96      0.95      0.95        15
weighted avg       0.94      0.93      0.93        15



### **Interpretación**

Como primera observación, el modelo KNN con datos escalados está funcionando mucho mejor, y eso tiene sentido:
**KNN mide distancias entre puntos** (usualmente con Euclídea), por lo que **si las variables no están en la misma escala, las que tienen valores más grandes dominan la distancia total.**


| Clase        | Precision | Recall | F1-score | Interpretación                                                                                          |
| ------------ | --------- | ------ | -------- | ------------------------------------------------------------------------------------------------------- |
| **apple**    | 0.80      | 1.00   | 0.89     | De todas las predicciones “apple”, el 80% fueron correctas, y se reconocieron todos los “apple” reales. |
| **lemon**    | 1.00      | 1.00   | 1.00     | Perfecto: el modelo clasificó todos los “lemon” correctamente.                                          |
| **mandarin** | 1.00      | 1.00   | 1.00     | Perfecto también, aunque hay pocas muestras (2).                                                        |
| **orange**   | 1.00      | 0.75   | 0.86     | Predijo bien todos los “orange” que clasificó, pero dejó pasar algunos (25% no detectados).             |  




In [34]:
# Resumen de las predicciones realizadas por el clasificador
print ("Matriz de confusión:\n", confusion_matrix(y_test, y_pred_scaled), "\n")

Matriz de confusión:
 [[4 0 0 0]
 [0 5 0 0]
 [0 0 2 0]
 [1 0 0 3]] 





               Predicción
              A   L   M   O
Real       ┌────────────────┐
Apple   →  | 4   0   0   0 |
Lemon   →  | 0   5   0   0 |
Mandarin→  | 0   0   2   0 |
Orange  →  | 1   0   0   3 |
           └────────────────┘

| Clase real   | Predicciones correctas | Predicciones incorrectas     | Interpretación                                           |
| ------------ | ---------------------- | ---------------------------- | -------------------------------------------------------- |
| **Apple**    | 4/4                    | 0                            | Todas las manzanas fueron clasificadas correctamente.  |
| **Lemon**    | 5/5                    | 0                            | Los limones también fueron clasificados perfectamente. |
| **Mandarin** | 2/2                    | 0                            | Las mandarinas fueron identificadas sin errores.       |
| **Orange**   | 3/4                    | 1 (clasificada como “apple”) | Un naranja fue confundido con una manzana.            |  

### **Conclusiones**

* El modelo identifica correctamente el 93% de las frutas.
* Todas las clases están bien representadas (gracias al stratify=y).
* El escalado fue clave para mejorar el rendimiento.
* Solo una confusión: una naranja que se parece demasiado a una manzana en los valores de altura, ancho o color_score.


### **Probando otro metodo de normalización = MinMaxScaler**


In [35]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

X_scaled = scaler.fit_transform(X)

# Dividir en entrenamiento y prueba con los datos escalados
X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y, random_state=42)

knn_scaled = KNeighborsClassifier(n_neighbors=3)
knn_scaled.fit(X_train, y_train)

y_pred_scaled = knn_scaled.predict(X_test)
print("\nReporte de clasificación:")
print(classification_report(y_test, y_pred_scaled))


Reporte de clasificación:
              precision    recall  f1-score   support

       apple       0.80      1.00      0.89         4
       lemon       1.00      1.00      1.00         5
    mandarin       1.00      1.00      1.00         2
      orange       1.00      0.75      0.86         4

    accuracy                           0.93        15
   macro avg       0.95      0.94      0.94        15
weighted avg       0.95      0.93      0.93        15



In [36]:
print ("Matriz de confusión:\n", confusion_matrix(y_test, y_pred_scaled), "\n")

Matriz de confusión:
 [[4 0 0 0]
 [0 5 0 0]
 [0 0 2 0]
 [1 0 0 3]] 



## **Interpretacion**
* Los resultados sonmuy similares al método anterior, pero, mucho mejor sin usar el escalamiento de los datos

## **Implementación del modelo para Predicciones sobre nuevas frutas (ficticios)**

In [38]:
# Especificamos las medidas de una fruta en una lista de valores
#  altura, ancho, masa, escala de color

fruta1 = [[5.5, 2.2, 150, 0.70]]
print("la fruta 1 con caraceristicas ", fruta1, "es de tipo",
     knn_scaled.predict(fruta1))

la fruta 1 con caraceristicas  [[5.5, 2.2, 150, 0.7]] es de tipo ['orange']


In [42]:
# 2 frutas
#  altura, ancho, masa, escala de color
frutas = [[5.5, 2.2, 120, 0.70], [2, 2, 160, 0.50 ]]
print("las frutas con caraceristicas ", frutas, "son de tipo:",
     knn_scaled.predict(frutas))

las frutas con caraceristicas  [[5.5, 2.2, 120, 0.7], [2, 2, 160, 0.5]] son de tipo: ['orange' 'orange']
